### PR TITLE
New module - iam_role

### PR DESCRIPTION
**Pull request by wimnat**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

iam_role
##### ANSIBLE VERSION

2.2
##### SUMMARY

Module to allow managing AWS IAM roles.  Similar functionality to current iam module but it was decided that the module should be split in to separate ones i.e. iam_role, iam_user, iam_group.

This module is the first of that split.

Extra functionality includes support for managed policies

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @wimnat for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by defionscode** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;X5PVjKHK&#x2F;12-iam-split-out"> iam split out</a>

**Comment by shaunbrady** [comment]
I would really like to see some new IAM modules.  I am having to work around some short comings of the original ones (iam, iam_policy).

**Comment by juliedavila** [comment]
ship_it

**Comment by shaunbrady** [comment]
@wimnat Are iam_user and iam_group in the works, coming from another source (core?), or future plans?  Thanks for the work.

**Comment by wimnat** [comment]
@shaunbrady iam_user is in the works.  I will hopefully get to iam_group too but might be a month or two

**Comment by shaunbrady** [comment]
@wimnat Perfect.  I'm probably look at a similar time frame, but if you want to point me at a branch, I might be able to toss some cycles at it (if you want help).

**Comment by naslanidis** [comment]
Hi @wimnat 

Does this module still support inline policies as well? 

It doesn't look like it does but correct me if I'm wrong. So I'm guessing the iam_policy module would still be used for that purpose and this new module for attaching managed policies?

Nick.

**Comment by wimnat** [comment]
@naslanidis you are correct.  You would create your role and attach any managed policies with this module and then attach any inline policies with the iam_policy module.

**Comment by a1max1** [comment]
what about creating managed policies?

**Comment by jingxiaogu1216** [comment]
@wimnat 
Hi! Could u give an example file of "policy.json"? I use your new module "iam_role".
This is my test.yml file
- hosts: localhost
  connection: local
  gather_facts: False
  
  tasks:
  - name: attach managed policies to iam role
    iam_role:
     name: mynewrole
    assume_role_policy_document: "{{ lookup('file','policy.json') }}"
    state: present
    managed_policy:
    - arn: aws:iam::aws:policy/ReadOnlyAccess  

This is my policy.json file
{
    "Statement": [
        {
            "Action": "s3:PutObject",
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "arn:aws:iam::127311923021:root",
                    "arn:aws:iam::027434742980:root",
                    "arn:aws:iam::797873946194:root",
                    "arn:aws:iam::156460612806:root",
                    "arn:aws:iam::054676820928:root",
                    "arn:aws:iam::582318560864:root",
                    "arn:aws:iam::600734575887:root",
                    "arn:aws:iam::114774131450:root",
                    "arn:aws:iam::783225319266:root",
                    "arn:aws:iam::718504428378:root",
                    "arn:aws:iam::507241528517:root"
                ]
            },
            "Resource": "arn:aws:s3:::atonns_ctg_test/*",
            "Sid": "AllowLogDeliveryFromOtherAcc"
        }
    ],
    "Version": "2012-10-17"
}

The error is:
fatal: [localhost]: FAILED! => {"changed": false, "error": {"code": "MalformedPolicyDocument", "message": "This policy contains invalid Json", "type": "Sender"}, "failed": true, "msg": "An error occurred (MalformedPolicyDocument) when calling the CreateRole operation: This policy contains invalid Json", "response_metadata": {"http_headers": {"content-length": "288", "content-type": "text/xml", "date": "Fri, 29 Jul 2016 09:32:32 GMT", "x-amzn-requestid": "60ac15c1-556f-11e6-8c6f-4db199007c25"}, "http_status_code": 400, "request_id": "60ac15c1-556f-11e6-8c6f-4db199007c25"}}

I was stuck for several hours.
Thanks!

**Comment by Constantin07** [comment]
@wimnat 

Despite the policy is optional it throws an error:

```
  - iam_role:
      name: Test_role
      path: "/test/"
      state: present
```

```
TASK [iam_role] ****************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "state is present but the following are missing: assume_role_policy_document"}
```

File: cloud/amazon/iam_role.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    assume_role_policy_document: "{{ lookup('file','policy.json') }}"
+    state: present
+
+# Create a role and attach a managed policy called "PowerUserAccess"
+- iam_role:
+    name: mynewrole
+    assume_role_policy_document: "{{ lookup('file','policy.json') }}"
+    state: present
+    managed_policy:
+      - arn:aws:iam::aws:policy/ReadOnlyAccess
```
Comment (shaunbrady) [inline cloud/amazon/iam_role.py]: Typo?  PowerUserAccess?

**Comment by wimnat** [comment]
@Constantin07 yeh i'm not sure about how to handle this.  It's required on a state:present but not a state:absent so not sure what to put in the doc.

@gregdek, @bcoca  - can you weigh in on the above?

**Comment by juliedavila** [comment]
@wimnat 

In the docstring, make assume_role_policy not required but make a note in the descritpion that it _is_ required when a state of present is supplied. The way I see it required should only be set to true when it is required 100% of the time.

**Comment by wimnat** [comment]
Thanks @defionscode.  Done.

**Comment by juliedavila** [comment]
@jingxiaogu1216 you need to do it as `" {{ lookup('file','policy.json') }}"` notice the space between `{{` and the "L" in lookup. It's a weird workaround to keep YAML from seeing the json as in inline dict.

**Comment by gregdek** [comment]
Thanks @wimnat for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by wimnat** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @wimnat for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by wimnat** [comment]
@Constantin07 @shaunbrady @naslanidis  you guys seem to of used this so if you think it's ok can you please add a shipit comment?

@ryansb - can we try and get this in a extras module meeting?  It's been around a while.

**Comment by Constantin07** [comment]
shipit

**Comment by shaunbrady** [comment]
@wimnat Sorry, I have not actually used the module, just wanting it's feature list.  I worked around the need for this for now, but it would greatly improve my playbooks.  I can try it out if there is further hold up.

**Comment by wimnat** [comment]
shipit

**Comment by padmanakate** [comment]
shipit

**Comment by PrasadDuvvi** [comment]
Hello,  Need Help!

I'm very new to Ansible.
I've just used **iam: ** module to create an IAM role and attached a **trust_policy:**  (something like ec2.amazonaws.com) 

How can I add a **managed_policy:** to my role something like below:
 managed_policy:
      - arn:aws:iam::aws:policy/PowerUserAccess

Is this currently supported by Ansible..   I've tried with module **iam_role:** but I got an error message as **_"ERROR! no action detected in task. This often indicates a misspelled module name, or incorrect module path."_**

Is the managed_policy currently being supported by Ansible ?   I can do this with CFN without any issues.

**Comment by a1max1** [comment]
not supported out of the box.  I had to build separate playbooks for that.

**Comment by wimnat** [comment]
@PrasadDuvvi - this isnt really the correct forum for such questions.  Try the ansible user google group or stack overflow.

The short answer is you just use iam_role module.  It sounds like perhaps your version of Ansible doesn't have it.  Check you have version 2.3.

Example:
```
- name: ECS Service IAM role
  iam_role:
    name: ecsServiceRole
    assume_role_policy_document: "{ \"Version\": \"2012-10-17\", \"Statement\": [{ \"Sid\": \"\", \"Effect\": \"Allow\", \"Principal\": { \"Service\": \"ecs.amazonaws.com\" }, \"Action\": \"sts:AssumeRole\" }] }"
    state: present
    managed_policy:
      - "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceRole"
```

**Pull request closed without merging**